### ISSUE TITLE
RFC: proposal to change instance allocation pattern for ABI flexability

### ISSUE THREAD

**Issue by vtjnash** [libuv/libuv#2630]
While thinking about issues we could address in v2, I was looking at the set of types we have that keep around `unused` fields or need to malloc some extra fields to hang off of it in order to make fixes and enhancements without breaking ABI (sizeof).

I'd like to request feedback on a proposal to make the libuv structs more opaque: instead of declaring the whole struct in the public header, have libuv only declare the public fields there. To allocate a struct, you must call the appropriate libuv function. For instance, instead of:
```c
uv_tty_t* tty = malloc(sizeof(uv_tty_t));
uv_tty_init(tty, UV_STDIN_FD);
```
You'd write:
```c
uv_tty_t* uv_tty_new(UV_STDIN_FD);
```
Where `uv_tty_t` is just:
```c
typedef struct uv_handle_s {
  void* data; /* user pointer */
  uv_loop_t* loop; /* read-only */
  uv_handle_type type; /* read-only */
} uv_handle_t;
typedef struct uv_handle_s uv_handle_t;
typedef struct uv_handle_s uv_stream_t;
typedef struct uv_handle_s uv_tty_t;
```

This would have been more constraining a few years ago, but now all of these public fields should have accessors, so completely opaque objects may be entirely acceptable.

And similar for `uv_req_t` (though a few `req_t` subtypes may expose some additional fields).

Transition strategy I think could also be pretty reasonable: For v1.x, we can introduce the new API as an optional usage. If you use the `uv_*_new` API, then `uv_close` will automatically free the memory after returning from the user callback. Otherwise, everything'd continue working as it does now. In v2.x, the old `uv_*_init` APIs would be removed, making it possible for libuv to then freely alter fields.

**Comment by addaleax** [comment]
Fwiw, this was mostly the idea behind 0d6525acae3b9 (and the earlier f09f7bc6a8b64585cd74d1fca712c6d0cab2f737) – allow ABI flexibility for those who want it. I’m good with formalizing that in some way, although I’m not sure that we should require an API switch like this…?

**Comment by vtjnash** [comment]
Right, if 0d6525a didn't already exist, it would have needed to be implemented as part of this. This would possibly eliminate the usefulness of f09f7bc (though it'd still be valid), since there'd no longer be anything particularly useful you could do from knowing the size. The current (v1) design exposes `sizeof` as an API detail, which constrains libuv from making changes which alter it.

**Comment by chrthi** [comment]
This would force users to use dynamic allocation per-handle. It is often useful to combine multiple uv_handles in a larger structure or place them in static memory. Removing that option would force me to migrate away from libuv, so please don't.

**Comment by vtjnash** [comment]
Replacing static memory with malloc is still `O(0)`. It sounds like you may be prematurely worried about an optimization that would make no observable difference. We can still keep the `uv_handle_size` as a runtime query, just static `sizeof()` would not be available. The `uv_malloc_func` is also already replaceable, so you can use a bump-pointer design (for example). We could also even consider expanding that to support integration with faster allocators, such as garbage collectors.

**Comment by bnoordhuis** [comment]
Using sized structs was one of the big design decisions for libuv and it's something I'd rather not change.

A pimpl-like approach, where the public struct is just a wrapper around a pointer that points to a heap-allocated internal implementation, might be a good middle ground.
```c
// include/uv.h
struct uv_imp_tty;

struct uv_tty_s {
  struct uv_imp_tty* imp;
};

typedef struct uv_tty_s uv_tty_t;

// src/unix/internal.h (or wherever)
struct uv_imp_tty {
  int fd;
  // ...
};
```
Things like the `data` field should perhaps still stay part of the public API.

pimpl makes FFI easier because all structs have uniform layout. (That's something we didn't pay too much attention to when designing libuv. Live and learn.)

**Comment by saghul** [comment]
Agreed with @bnoordhuis. That should also help with making a "forever vX" release I guess, since the potential for ABI breakage would be heavily reduced and we could always make changes in the private struct.

**Comment by vtjnash** [comment]
I guess the idea is that `typedef struct uv_tty_s uv_tty_t;` is more type-safe than what I posted? I don't particularly like the idea of passing around `uv_handle_t` as a first class aggregate (awkward for FFI, sometimes has a slow calling convention, and requires changing a lot more code to drop the `*` on the type decls and replace `->data` with `.imp->data`. Passing around `uv_handle_t*` with the pimpl design is also a non-starter for me, as the point of this issue was that we can do this without generally altering the typical allocation patterns I've usually seen. Whereas to me, pimpl ruins the original point of using sized structs (of knowing that libuv won't allocate any extra memory and gets everything with one pointer indirection).

Though I'd perhaps add a C99 unsized array to the struct decl I wrote above to break `sizeof` at compile-time though, to help catch errors if the user isn't just using the pointer itself (e.g. `uv_handle_t*`):
```
typedef struct uv_handle_s {
  /* public fields */
  void* data; /* user pointer */
  uv_loop_t* loop; /* read-only */
  uv_handle_type type; /* read-only */
  /* private fields */
  int _disable_sizeof_[];
} uv_handle_t;
```

**Comment by bnoordhuis** [comment]
An important aspect of sized structs is that they allow embedding, i.e., they allow container_of-like macros to work. That stops working with your proposal, if I understand it correctly.

**Comment by vtjnash** [comment]
Yep, but I assumed the point of that is to allow avoiding the pimpl design and thus the excess allocation. If libuv is already doing an allocation, then I think that `uv_tty_t*` is superior to `struct { uv_imp_tty_t* imp; }`, given that both have the same `sizeof` and embedding characteristics.

> That stops working with your proposal, if I understand it correctly

It changes, but isn't eliminated. There's a couple options to keep it working that I've used in the past:
1. `uv_tty_new` could take a `sizeof` argument for additional bytes to hang off the end (or beginning) and put that pointer in `->data` on return
2. `uv_malloc_func` can be extended to accept additional context information, as I mentioned above, permitting the use of placement new (aka struct embedding) and other allocation strategies. This could be either just the libuv `type` or an entire `void*` context argument to the `uv_new` functions.
3. The libuv API could continue to provide `uv_tty_init`, but simply stop providing the compile-time computation of the sizeof. This makes the struct slightly more annoying to use in embedding, since the offsets (for `container_of` and `malloc`) become expressions instead of constants, but otherwise works about the same as now.

We can do any one, or all three, so it's less about tradeoffs between them and which ones are in demand to implement.

**Comment by saghul** [comment]
> Passing around `uv_handle_t*` with the pimpl design is also a non-starter for me, as the point of this issue was that we can do this without generally altering the typical allocation patterns I've usually seen. Whereas to me, pimpl ruins the original point of using sized structs (of knowing that libuv won't allocate any extra memory and gets everything with one pointer indirection).

Why is that so bad? Is it your preference, are you concerned about performance, something else?

**Comment by vtjnash** [comment]
Yeah. We might be talking past each other, but doing everything via a double pointer indirection seems inconvenient both for code style and performance.

Note that v1.x is already in the situation now that "force users to use dynamic allocation per-handle" is happening due to running out of room in the existing structs (e.g. https://github.com/libuv/libuv/pull/2725, https://github.com/libuv/libuv/pull/2169, https://github.com/libuv/libuv/pull/790), which is what this proposal is intended to address for v2 (ensuring libuv always does one allocation and the caller does zero, instead of each sometimes doing one).
